### PR TITLE
WIP: feat(dsl): support a map of structs in matcher

### PR DESCRIPTION
**Pull request by jfkriz**
Example JSON which this matcher will support:
```
{
  "departments": {
    "10": {
      "name": "Men's Clothing",
      "hours": "10am - 9pm"
    },
    "20": {
      "name": "Women's Clothing",
      "hours": "10am - 9pm"
    }
  }
}
```

### CODE REVIEW

**Comment by mefellows** [comment]
Thanks @jfkriz, i'm open to this PR. If you can please update and ensure the tests are working i'll bring this in.

**Comment by jfkriz** [comment]
Thanks Matt.  I'm working on unit tests - I was quickly doing this on a
machine that is not my primary dev machine, and did not have Go properly
setup, and just missed the unit tests for this.  I'm getting things setup
now, and have test coverage in progress - will update the PR and remove the
WIP status when ready.  On a somewhat related note, do you typically like
to see changes merged into all current branches/releases, or is it ok to
just get it working as a new feature in the 1.x.x release?

Thank you - and apologies for the broken build and premature PR on this!

**Comment by mefellows** [comment]
Thanks @jfkriz. The `1.x.x` branch is appropriate, I'll merge into `master` when its stable etc.

> Thank you - and apologies for the broken build and premature PR on this!

No worries at all, i'd rather you do this and make your intentions clear rather than work away on it for days just to have me say "It's not coming in because of X/Y/Z" .

So in terms of getting this in, if I understood correctlp, the intention is that any map of shape `map[string]<struct>` should be able to be provided and automatically converted into a valid Pact matcher?

If so there are a few things to consider:

1. ``Like(`"string"`)`` which you've re-used is actually as it turns out a minor bug, it should simply be `Like("string")`
2. Do we need to consider arbitrary nesting in the test cases?
3. If you could please update the example in `examples/consumer_test.go` to demonstrate compatibility?

i.e. this block:

```go
			Body: map[string]string{
				"name": "billy",
			},
```

should be able to be replaced like-for-like and produce the same pact file (except with extra matchers):

```go
			Body: dsl.Match(map[string]string{
				"name": "billy",
			}),
```

**Comment by mefellows** [comment]
Sorry, accidentally hit close on mobile.

I was thinking about the PR this morning and thought I Could help byputting in some example tests for this feature. I think the current ones you've provided will produce invalid matching rules and won't work on the provider side.

Hopefully i can look this afternoon.

File: dsl/matcher_test.go [hunk 1]
```diff
@@ -694,9 +694,40 @@ func TestMatch(t *testing.T) {
...
+			name: "map[string]int",
 			args: args{
-				src: make(map[string]string),
+				src: make(map[string]int),
+			},
+			want: map[string]interface{}{
+				`"string"`: Like(1),
+			},
+		},
+		{
```
Comment (mefellows) [inline dsl/matcher_test.go]: I think this should be:

```go
		{
			name: "map[string]struct",
			args: args{
				src: map[string]wordDTO{
					"word1": wordDTO{},
					"word2": wordDTO{},
				},
			},
			want: map[string]interface{}{
				"word1": Matcher{
					"word":   Like("string"),
					"length": Like(1),
				},
				"word2": Matcher{
					"word":   Like("string"),
					"length": Like(1),
				},
			},
		},
```

At the moment, you lose the second k/v pair

**Comment by mefellows** [comment]
OK the more I think on this, the more I'm not sure it makes sense to implement - but I want to hear more from your use case.

Apart from the specific use-case, where there are well-known structs (let's call your example a `Department`) that are wrapped inside another object and keyed by strings(`"10"`, `"20"`), all existing matchers should do the job - either a `dsl.Like(<some interface>)` or a `dsl.Match(<some struct>)`) would produce relatively sane defaults.

For your example, I would generally encourage clients to create a wrapper type for this object (e.g. `DepartmentResults`), and then you could use the field tags of `dsl.Match` for specific validation needs. If you just cared about the types, then you could add it to `dsl.Like` as is, and it would create expectations where all of the keys must exist, but the values just need to match the type.

In your case, my suspicion is that the "10" and "20" can be dynamic, and at the moment we don't support dynamic keys (see https://github.com/pact-foundation/pact-specification/issues/47 and https://github.com/pact-foundation/pact-specification/blob/version-4/README.md#ignoring-the-keys-in-a-map) but we want to introduce it (we have only implement v2 of the specification and have parts of v3 in flight). We currently have a dependency on https://github.com/pact-foundation/pact-ruby-standalone for the underlying validation engine before it will flow into this library.

Back to your example, assuming we want to bring this in, your [matcher_test.go](https://github.com/pact-foundation/pact-go/pull/91/files#diff-cf0e1f508a4d132748c2d18f87391964R705) example should look like this:

```
		{
			name: "map[string]struct",
			args: args{
				src: map[string]wordDTO{
					"word1": wordDTO{},
					"word2": wordDTO{},
				},
			},
			want: map[string]interface{}{
				"word1": Matcher{
					"word":   Like("string"),
					"length": Like(1),
				},
				"word2": Matcher{
					"word":   Like("string"),
					"length": Like(1),
				},
			},
		},
```

Otherwise we'll lose one of the keyed items, and the keys will be incorrect. When this gets verified on the provider side, the keys must match exactly (even if the values don't).

If on the other hand we could somehow address `(2)` above (nested maps with tagged structs) this would be a very useful feature as it is much more flexible.

Dose this make sense?

**Comment by mefellows** [comment]
cc: @bethesque

**Comment by jfkriz** [comment]
Matt, thank you so much for the information above.  The issue and the README link in particular are describing almost exactly what we're trying to do (and your suspicion is correct on that as well).  I am relatively new to Pact, and we're trying to use it to validate some services we are integrating with, and unfortunately, we don't have full control over those service contracts.  In our case, the key is dynamic, and we don't care so much about the value of the key, but rather that the tagged struct referenced by the key can be verified.  As I was working through this yesterday, I was starting to come to the conclusion that this was not currently supported in the validation engine, and you've confirmed that for me.  I apologize for missing this in the first place - I did look through issues on pact-go, but did not look through issues on the pact-specification, so that was my miss.

Do you think it will be possible to solve this (nested maps with tagged structs) without any changes in the validation engine?  If so, I am happy to continue working on this code to get to something that will be useful for my team, and for the larger Pact community.

Thanks again, you've been very helpful!

**Comment by jfkriz** [comment]
Thanks Matt.  I would really like the dynamic keys part, as that's what I
initially was trying to accomplish with this PR.  Admittedly, it's been a
while (early 2000s) since I've done anything significant in Ruby, but I'm
willing to dive back in if I can help move that effort forward.

**Comment by mefellows** [comment]
@bethesque would you be able to provide some pointers to @jfkriz on how/where dynamic keys might be able to be added to the Ruby engine?

**Comment by mefellows** [comment]
Also, I'm going to close this for now. I imagine a new PR will be required for the dynamic keys discussion.

**Comment by bethesque** [comment]
> Do you think it will be possible to solve this (nested maps with tagged structs) without any changes in the validation engine?

Unfortunately not.

Dynamic keys aren't even implemented in the v3 Rust impl. They won't be introduced until v4 unfortunately. https://github.com/pact-foundation/pact-specification/issues/47

**Comment by mefellows** [comment]
@bethesque I think @jfkriz was willing to help implement dynamic keys in Ruby, any tips/advice?

**Comment by bethesque** [comment]
I'm not sure about introducing v4 features that we don't even have a pact specification for, before we've implemented v3 ones. We'd have to work out with @uglyog how we'd represent that in the pact specification.

I'm actually not even sure how the matching logic would work. If, for instance, the key had to match `/\d/` and had an example value given, do we verify that every key that matches `/\d+/` has the body, or only one? Or do we allow the user to specify either way? How do we represent this in the matching rules? These are the sort of things that need to be thrashed out before we can even touch the code.

**Comment by jfkriz** [comment]
Thanks for the info @bethesque.  Whenever v4 work for this feature begins (including the hashing out of the details on what the matching logic should be), I am happy to lend a hand, and can offer my input as a potential user of this feature as well.  And thanks to both you and @mefellows for being responsive and helpful to both of my issues.  I'm new to Pact, but can already see the value it will bring to our organization, and I'm pretty excited to really dig into it and help out the Pact community wherever I can along the way.

**Comment by bethesque** [comment]
No worries @jfkriz. If you're keen on getting involved, would you like to get your head around the way the existing matching rules are serialised, and come up with a proposal for how the key matching could be serialised? I'm pretty sure @uglyog hasn't started on this.

The wip version of v4 is here https://github.com/pact-foundation/pact-specification/tree/version-4/testcases/request/body

**Comment by jfkriz** [comment]
@bethesque , yes, I'll take a look at that, get a better understanding of how all existing rules are serialized, and come up with a proposal (or maybe several options) for what this might look like.  To clarify, based on what [the v4 specification](https://github.com/pact-foundation/pact-specification/tree/version-4#ignoring-the-keys-in-a-map) currently says, the goal would be to ignore the key, but verify the value.  However, in [your earlier comment](https://github.com/pact-foundation/pact-go/pull/91#issuecomment-407238055) on this PR, you mentioned matching keys.  I am personally in favor of this approach, as I'd like to match keys in much the same way I could match a value - but I think it would be useful to offer this as an option when generating the pact.  Then I can opt to completely ignore the key value, and just make sure the value matches, or I can opt to validate that the key matches a pattern as well.

**Comment by bethesque** [comment]
I think we'd want the same options for the keys as we have for the values - we'd want to say, it can be any string (which I think is effectively the same as ignoring the value, as you mentioned), or it can match this regular expression.

**Pull request closed without merging**